platform/keysystem/keyclient/actloop/actloop.go

```diff
@@ -20,45 +20,63 @@ type Action interface {
 	Info() string
 }
 
-func NewActLoop(actions []Action, logger *log.Logger) ActLoop {
-	return ActLoop{actions: actions, logger: logger}
+type NewAction func(nac *NewActionContext)
+
+type NewActionContext struct {
+	Logger    *log.Logger
+	BlockedBy []error
+	Performed bool
+}
+
+func (nac *NewActionContext) Errored(info string, err error) {
+	nac.Logger.Printf("action stop error: %s (in %s)\n", err.Error(), info)
+}
+
+func (nac *NewActionContext) Blocked(err error) {
+	nac.BlockedBy = append(nac.BlockedBy, err)
+}
+
+func (nac *NewActionContext) NotifyPerformed(info string) {
+	nac.Logger.Printf("action performed: %s\n", info)
+	nac.Performed = true
 }
 
-func (m *ActLoop) Step() (stabilized bool, blocked bool) {
-	var blockedBy []error
-	for _, action := range m.actions {
+func ActionToNew(action Action) NewAction {
+	return func(nac *NewActionContext) {
 		pending, err := action.Pending()
 		if err != nil {
-			m.logger.Printf("actloop check error: %s (in %s)\n", err.Error(), action.Info())
-		}
-		if !pending {
-			continue
-		}
-		blockerr := action.CheckBlocker()
-		if blockerr != nil {
-			blockedBy = append(blockedBy, blockerr)
-			continue
-		}
-		err = action.Perform(m.logger)
-		if err != nil {
-			m.logger.Printf("actloop step error: %s (in %s)\n", err.Error(), action.Info())
-		} else {
-			m.logger.Printf("action performed: %s\n", action.Info())
-			return false, false
+			nac.Errored(action.Info(), err)
+		} else if pending {
+			blockerr := action.CheckBlocker()
+			if blockerr != nil {
+				nac.Blocked(blockerr)
+			} else {
+				err = action.Perform(nac.Logger)
+				if err != nil {
+					nac.Errored(action.Info(), err)
+				} else {
+					nac.NotifyPerformed(action.Info())
+				}
+			}
```

Comment:
I have a strange feeling that we're implementing some weird version of promises.  Can we discuss the action loop design again?  It's been 1.5 years since the last time you explained to me why we need it.

Reply:
The actloop is really just an abstraction to facilitate converging a bunch of pieces of state. Each action represents something that could need to be done, and (in the previous design) had functions to:
 - check whether the action needs to be done (Pending)
 - check whether there are any unsatisfied preconditions for the action (CheckBlocker)
 - attempt to perform the action (Perform)

The actloop keeps attempting to perform actions until there are no actions left that can be performed. Note that actions can go from a "done" state to a "needs to be done again" state, such as if a certificate is getting close to expiring, in which case it will be re-run. So I don't think it's quite like promises.